lib/phoenix_html/engine.ex

```diff
@@ -16,7 +16,39 @@ defmodule Phoenix.HTML.Engine do
   def init(_opts), do: {:safe, ""}
 
   @doc false
-  def handle_body(body), do: body
+  def handle_body(body) do
+    compact_safe(body)
+  end
+
+  defp compact_safe({:safe, body}) when is_list(body) do
+    case compact_static(List.flatten(body)) do
+      [element] ->
+        {:safe, element}
+      list ->
+        {:safe, list}
+    end
+  end
+  defp compact_safe({:safe, other}) do
+    {:safe, other}
+  end
+  defp compact_safe(other) do
+    other
+  end
+
+  defp compact_static([binary1, binary2 | rest])
+       when is_binary(binary1) and is_binary(binary2) do
+    compact_static([binary1 <> binary2 | rest])
+  end
+  defp compact_static([{_, _, _} = ast | rest]) do
+    compacted = Macro.prewalk(ast, &compact_safe/1)
```

Comment:
Can you please explore compacting as we build the AST? That would ease my mind. Unfortunately anyone can create a `{:safe, ...}` tuple.